### PR TITLE
Run app automatically and generate mainModuleId

### PR DESCRIPTION
**Pull request by adrw**
- Boot app after initializer, add manifest to initializer
- Auto-generate mainModuleId from last sorted module

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -71,13 +71,7 @@ must be confined to a single thread.
...
   val manifestUrl = "http://localhost:8080/manifest.zipline.json"
   val loader = ZiplineLoader(dispatcher, OkHttpClient())
-  return loader.loadOrFail("trivia", manifestUrl) {
-    val moduleName = "./zipline-root-trivia-js.js"
-    it.quickJs.evaluate(
-      "require('$moduleName').app.cash.zipline.samples.trivia.launchZipline()",
-      "launchZiplineJvm.kt"
-    )
-  }
+  return loader.loadOrFail("trivia", manifestUrl)
```
Comment (swankjesse) [inline README.md:74]: YESSSSS

File: zipline-gradle-plugin/README.md [hunk 2]
```diff
@@ -8,9 +8,12 @@ The plugin compiles JavaScript files into `.zipline` files. These binary files a
 
 The plugin requires setting the following variables which are used in compilation.
 
-- `mainModuleId`: This is the JS module that your application's entrypoint function is in. This will usually be the last key in the `modules` map in the generated ZiplineManifest which is sorted topologically.
 - `mainFunction`: This is the fully qualified function call to start your application, note it does include the trailing `()` to initiate the call.
```
Comment (swankjesse) [inline zipline-gradle-plugin/README.md:12]: I missed this in my last PR. . . . change the word `requires`  to `offers` on line 9 ?

File: zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplineCompileTask.kt [hunk 3]
```diff
@@ -51,7 +51,7 @@ abstract class ZiplineCompileTask @Inject constructor(
   val outputDir: DirectoryProperty = objectFactory.directoryProperty()
 
   @get:Input
-  val mainModuleId: Property<String> = objectFactory.property(String::class.java)
+  val mainModuleId: Property<String?> = objectFactory.property(String::class.java)
```
Comment (swankjesse) [inline zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplineCompileTask.kt]: ty

File: zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplineCompileTask.kt [hunk 4]
```diff
@@ -51,7 +51,7 @@ abstract class ZiplineCompileTask @Inject constructor(
...
 
   @get:Input
   val mainFunction: Property<String> = objectFactory.property(String::class.java)
```
Comment (swankjesse) [inline zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplineCompileTask.kt:63]: String?

File: zipline-gradle-plugin/README.md (hunk 2 above)
Comment (swankjesse) [reply zipline-gradle-plugin/README.md:12]: Oh drat my PR isn’t merged. Rebase on that??

File: zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineLoader.kt [hunk 5]
```diff
@@ -188,7 +189,12 @@ class ZiplineLoader internal constructor(
       receive(ZiplineLoadReceiver(zipline), manifest, applicationName)
 
       // Run caller lambda to validate and initialize the loaded code to confirm it works.
-      initializer(zipline)
+      initializer(zipline, manifest)
+
+      // Run the application after initializer has been run on Zipline engine.
+      if (runAfterInitializer) {
```
Comment (swankjesse) [inline zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineLoader.kt]: For follow-up, I wonder if we can make this always true? Or maybe, always true if `mainFunction` is not null

File: zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineManifest.kt [hunk 6]
```diff
@@ -46,23 +46,25 @@ data class ZiplineManifest private constructor(
...
         }
-        .associateWith { id ->
+      return ZiplineManifest(
+        modules = sortedModuleIds.associateWith { id ->
           modules[id]
             ?: throw IllegalArgumentException("Unexpected [id=$id] is not found in modules keys")
-        }
-    )
+        },
+        mainModuleId = mainModuleId ?: sortedModuleIds.last(),
```
Comment (swankjesse) [inline zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineManifest.kt:67]: NICE

File: zipline-loader/src/commonTest/kotlin/app/cash/zipline/loader/ZiplineManifestTest.kt [hunk 7]
```diff
@@ -136,11 +130,29 @@ class ZiplineManifestTest {
     )
   }
 
+  @Test
+  fun usesLastSortedModuleAsMainModuleId() {
```
Comment (swankjesse) [inline zipline-loader/src/commonTest/kotlin/app/cash/zipline/loader/ZiplineManifestTest.kt:139]: 👍🏻

**Review by swankjesse — APPROVED** [review]

File: zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineLoader.kt (hunk 5 above)
Comment (adrw) [reply zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineLoader.kt]: I didn't know if there'd be a case where we don't want the app immediately booted so I left it configurable. If that's not a case we'd ever want, happy to simplify and do it by default.

Comment (swankjesse) [reply zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineLoader.kt]: I think we make it on-always for now and add configurability later if we need it.

**Review by swankjesse — APPROVED** [review]

File: zipline/src/engineMain/kotlin/app/cash/zipline/Zipline.kt [hunk 8]
```diff
@@ -165,6 +165,11 @@ actual class Zipline private constructor(
     quickJs.evaluate("delete globalThis.$CURRENT_MODULE_ID;")
   }
 
+  /** Runs the entrypoint to launch the application. */
```
Comment (swankjesse) [inline zipline/src/engineMain/kotlin/app/cash/zipline/Zipline.kt]: Can you move this into a private function in ZiplineLoader? I don’t think it’s necessary as a public API here

**Review by swankjesse — APPROVED** [review]

**Comment by adrw** [comment]
@swankjesse I'm getting some local build failures for undefined Zipline/QuickJS call and perpetual builds for Android/Native in CI. Any ideas?

**Review by swankjesse — APPROVED** [review]

**Comment by swankjesse** [comment]
Might just be a test issue? I added `e.printStackTrace()` in the catch clause of `loadOrFallBack()` and it gave me this:

```
app.cash.zipline.QuickJsException: "apple" not found in []
	at JavaScript.<anonymous>(define.js:11)
	at JavaScript.<eval>(zipline-main.js)
	at app.cash.zipline.QuickJs.evaluate(Native Method)
	at app.cash.zipline.QuickJs.evaluate(QuickJs.kt:110)
	at app.cash.zipline.loader.ZiplineLoader.runMainFunction(ZiplineLoader.kt:172)
	at app.cash.zipline.loader.ZiplineLoader.createZiplineAndLoad(ZiplineLoader.kt:200)
	at app.cash.zipline.loader.ZiplineLoader.access$createZiplineAndLoad(ZiplineLoader.kt:62)
	at app.cash.zipline.loader.ZiplineLoader$createZiplineAndLoad$1.invokeSuspend(ZiplineLoader.kt)
```

**Comment by swankjesse** [comment]
... I think we’re incorrectly attempting to load the module, and our test fixtures don’t actually define any modules.

Fix might be to change this thing to export a module? Or to change the generated manifest to not use a module.
```
    fun createJs(seed: String) = """
      |globalThis.log = globalThis.log || "";
      |globalThis.log += "$seed loaded\n"
      |""".trimMargin()
```

**Comment by adrw** [comment]
Not sure why some of the builds are timing out

**Comment by adrw** [comment]
I'm seeing local failures of the `ZiplinePluginTest.endToEnd` test with the following.

```
Exception in thread "main" java.lang.IllegalStateException: A global JavaScript object called app_cash_zipline_inboundChannel was not found
	at app.cash.zipline.QuickJs.getInboundCallChannel(Native Method)
	at app.cash.zipline.QuickJs.getInboundChannel$zipline(QuickJs.kt:118)
	at app.cash.zipline.Zipline$endpoint$1$jsInboundBridge$2.invoke(Zipline.kt:60)
	at app.cash.zipline.Zipline$endpoint$1$jsInboundBridge$2.invoke(Zipline.kt:59)
	at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81)
	at app.cash.zipline.Zipline$endpoint$1.getJsInboundBridge(Zipline.kt:59)
	at app.cash.zipline.Zipline$endpoint$1.call(Zipline.kt:69)
	at app.cash.zipline.internal.bridge.OutboundCallHandler.call(OutboundCallHandler.kt:57)
	at app.cash.zipline.tests.GreetService$Companion$Adapter$GeneratedOutboundService.greet(GreetService.kt:20)
	at app.cash.zipline.tests.LaunchGreetServiceJvmKt$main$1.invokeSuspend(launchGreetServiceJvm.kt:56)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829
```

File: zipline-loader-testing/src/engineMain/kotlin/app/cash/zipline/loader/testing/LoaderTestFixtures.kt [hunk 9]
```diff
@@ -138,13 +136,52 @@ class LoaderTestFixtures {
...
+            """.trimIndent()
+    )
+
+    fun createFailureJs(seed: String) = jsBoilerplate(
+      seed = seed,
+      loadBody = "throw Error('$seed');",
+      mainBody = "throw Error('$seed');"
+    )
+
+    private fun jsBoilerplate(seed: String, loadBody: String, mainBody: String) = """
```
Comment (swankjesse) [inline zipline-loader-testing/src/engineMain/kotlin/app/cash/zipline/loader/testing/LoaderTestFixtures.kt:157]: This is awesome

**Pull request merged**